Change graphql library to use Strawberry and fix memory leak with large numbers subscription

rjwills28 opened the pull request:
This PR includes the following major changes:

- We have changed the graphql library from Tartiflette to Strawberry. 
- We have also fixed a memory leak within Coniql which occurs with many subscriptions (~200) to PVs updating at 10Hz. This was found when running performance tests stressing the application. This was due to Coniql not processing the graphql results quickly enough and so an internal queue continued to grow leading to ever increasing memory and out-of-date PV values. To fix this I removed the queue and added a deque. This means that only the most recent value with be sent and if we fall behind we simply have to skip that update. 

Other changes after a review of PR #39:

- Remove the 'resolver' layer which was an implementation necessary for Tartiflette.
- Remove creation of the 'context' and instead use a global value in the app.

Changes to tests:

- Move the IOC required for the tests to conftest.py.
- Add tests covering both websocket protocols (legacy and new).
- Use a single running IOC for all tests so we don't have to build and tear down in between each test.

AlexanderWells-diamond reviewed (changes requested):
I like this overall. I definitely like the code schema rather than the separate `.gql` file that `tartiflette` used. 

I note that `test_pvaplugin.py` seems to have not been updated - what is the intention with that file? I note that `pvaplugin.py` was updated so I expected the tests to be updated.

A number of my comments are duplicates of each other, where i've noted the same issue appearing in multiple places. There's also a bunch here that are by no means musts - a lot are optional or simply opinions.

aawdls commented on src/coniql/app.py:
```diff
@@ -1,49 +1,75 @@
-        error_coercer=error_coercer,
-        modules=["coniql.resolvers"],
+    # Create the GraphQL view to attach to the app
+    view = GraphQLView(
+        schema=strawberry_schema,
+        subscription_protocols=[GRAPHQL_TRANSPORT_WS_PROTOCOL, GRAPHQL_WS_PROTOCOL],
```
I agree that:
- It should be off for production
- It should be possible to enable it for testing

rjwills28 replied:
If you run locally, try going to `http://localhost:8080/ws`. This is the end point that we configure in the application and is where you can access the GraphIQL interface.

AlexanderWells-diamond replied:
Thanks, I can access it from that URL.

rjwills28 replied:
I've added a command line argument `--graphiql` which will enable the GraphiQL interface, but by default (and so for production) it will be disabled. Commit https://github.com/dls-controls/coniql/pull/44/commits/eaa0f347e991116dbd1508b30dc97d5fa6114784.

aawdls commented on tests/test_caplugin.py:
```diff
@@ -1,355 +1,167 @@
     assert len(results) == 3
 
 
+# !! Must be the last test as it calls a disconnect
```
IOC lifecycle management in a test environment is tricky because the processes seem to not always be cleaned up well.

@MJGaughran having worked recently on the topup integration tests, do you have any suggestions for the pattern to use for that?

rjwills28 replied:
I agree that it is not very nice having to maintain an order to the tests and this would likely lead to maintainability issues so thank you for catching this.
I have extracted some of the methods used to create the session IOC so that I can call them manually to run a separate IOC for just the one test (i.e. the one that tests a disconnect). I purposely changed the test ordering to prove that this works. Please see commit: [69305fb](https://github.com/dls-controls/coniql/pull/44/commits/69305fbf12c553e6c5f97b95f9caafbfb4d33c0b).

aawdls commented:
Thanks for looking at the dependencies to work with 3.7

In the future we do not need to maintain 3.7 compatibility, as we will not release this via the Diamond RHEL7 build server but deploy it in a container. So we can support any python version and move to stating a compatible range.

After this branch is merged we will apply this template https://github.com/diamondlightsource/python3-pip-skeleton which should enable e.g. testing against a matrix of verisons.

rjwills28 commented on Pipfile:
```diff
@@ -8,10 +8,12 @@ coniql = {editable = true, extras = ["dev"], path = "."}
 lint = "pre-commit run --all-files --show-diff-on-failure --color=always -v"
-tests = "pytest"
+# Not running tests until full subscription/query/put are implemented
+# tests = "pytest"
```
The only reason I hadn't put the automated tests back in was because there was still a test for the pvacces plugin which has not been updated as we deemed this not a current priority. However I agree that the other tests are now ready to run and so I have removed the pvaccess tests as we can add these when we add the support for pvaccess. See commit: [a14ace5](https://github.com/dls-controls/coniql/pull/44/commits/a14ace5f57f893fa844479d1e61bdf390ef6f650).

rjwills28 commented on src/coniql/pvaplugin.py:
```diff
@@ -5,7 +5,8 @@
 from coniql.coniql_schema import DisplayForm, Widget
-from coniql.device_config import ChannelConfig
+
+# from coniql.device_config import ChannelConfig
```
The pvaplugin has not been updated yet and I'm not sure was working properly to begin with so I will remove this file and recreate it when we add pva support.

rjwills28 replied:
Removed file - defer implementing pvaccess

rjwills28 commented on tests/conftest.py:
```diff
@@ -1,16 +1,249 @@
+@pytest.fixture(scope="session")
+def list_put():
+    return (
+        r"""
```
Good catch - there's no need for this to be an r-string. Removed in commit [b42ef84](https://github.com/dls-controls/coniql/pull/44/commits/b42ef84441aaf10ef944b3c03471bbe63d9a6574).

rjwills28 commented on tests/conftest.py:
```diff
@@ -1,16 +1,249 @@
+@pytest.fixture(scope="session")
+def base64_put():
+    value = json.dumps(json.dumps(BASE64_0_1688_2))
+    return r"""
```
Removed in commit [b42ef84](https://github.com/dls-controls/coniql/pull/44/commits/b42ef84441aaf10ef944b3c03471bbe63d9a6574).

rjwills28 commented on src/coniql/caplugin.py:
```diff
@@ -214,19 +252,57 @@ async def subscribe_channel(
             while True:
-                update = await q.get()
-                yield maker.channel_from_update(**update)
+                await asyncio.sleep(0.01)
```
We had an issue where the processing of subscription results could not keep up and so a queue continued to grow leading to a memory leak (see https://github.com/dls-controls/coniql/issues/38). I followed some of Tom's suggestion in the comments to use deques and signals on when a value is ready to be processed. This removed any object that we can `await` on and is the reason for this change.

I was originally using `asyncio.sleep(0)`, which is used to suspend the current coroutine and allows the event loops to run other coroutines than have been suspended. However I found that this was terrible for performance of Coniql. The CPU was constantly at 100% and updates were missed as a result. I did some fine tuning with the performance tests and found that 0.01s was the optimum wait time. Any shorter and I saw the CPU shoot up while any longer we dropped more updates.

rjwills28 commented on src/coniql/strawberry_schema.py:
```diff
@@ -0,0 +1,370 @@
+    """
+
+    @strawberry.field
+    async def float(root) -> Optional[float]:
```
Yes you are right, this is very confusing. I have spent a lot of time trying to fix this issue and make it clearer to the user. In the end I had to change the way we call the resolvers. Initially I was using the @strawberry.field annotation to define the resolver in line with the schema but this does not play well with mypy and I couldn't find any way to fix the type errors. So I have now adopted the other method the strawberry provide which is to use the field definition to define the resolver method to call. This is much more compliant with mypy.
e.g. value: float = strawberry.field(resolver=function_to_get_my_value

Along with this I have started to try and merge some of the repeated dataclasses between the strawberry schema definition and the 'types'. More to do.

Added type hints and mypy integration in commit: [7f137c1](https://github.com/dls-controls/coniql/pull/44/commits/7f137c183012c7b3d79627de8d9beec6a023b423)

rjwills28 commented on src/coniql/pvaplugin.py:
```diff
@@ -140,7 +139,7 @@ def get_display(self) -> Optional[ChannelDisplay]:
                     display.units,
                 ) = self.get_form_precision_units()
-            update_display(display, self.config)
+            # update_display(display)
```
Removed file - defer implementing pvaccess

rjwills28 commented on src/coniql/pvaplugin.py:
```diff
@@ -170,7 +169,7 @@ def get_display(self) -> Optional[ChannelDisplay]:
                 choices=v_value.choices,
             )
-            update_display(display, self.config)
+            # update_display(display)
```
Removed file - defer implementing pvaccess

rjwills28 commented on src/coniql/strawberry_schema.py:
```diff
@@ -0,0 +1,370 @@
+        # Remove the transport prefix from the read pv
+        pv = store.transport_pv(id)[1]
+        async for channel in plugin.subscribe_channel(pv):
+            yield SubscribeChannel(channel_id, channel)
```
Thanks, yes I saw these too. I have fixed this in my latest commit to be mypy compliant ([7f137c1](https://github.com/dls-controls/coniql/pull/44/commits/7f137c183012c7b3d79627de8d9beec6a023b423)).

rjwills28 commented on src/coniql/pvaplugin.py:
```diff
@@ -68,6 +67,7 @@ def update_display(display: ChannelDisplay, config: ChannelConfig):
     display.description = config.description or display.description
     display.widget = config.widget or display.widget
     display.form = config.display_form or display.form
+"""
```
Removed file - defer implementing pvaccess

rjwills28 commented on src/coniql/app.py:
```diff
@@ -1,49 +1,75 @@
+        execution_context: Optional[ExecutionContext] = None,
+    ) -> None:
+        for error in errors:
+            print(error.message)
```
I tested out the mechanism described in the link and unfortunately it doesn't do what we want it to do. It allows you to intercept the graphql result and any errors before you send it over the websocket, which means that you can manipulate the error sent to the client. However it doesn't have any affect on what is logged as part of the Coniql application, i.e. the stack trace always gets printed in the Coniql terminal.

The implementation I currently have is the only way I can find to give us the option to manipulate what is actually logged and hide the stack trace in the deployed system.

AlexanderWells-diamond replied:
Fair enough. I'd like to see a note that this is undocumented, and therefore may break with Strawberry updates without notice. Just so the next developer has some idea why it may suddenly be broken.

rjwills28 replied:
Added a comment about this in commit: https://github.com/dls-controls/coniql/pull/44/commits/21986ea61570d37ee21a8529aa645f70623506c9

rjwills28 commented on src/coniql/app.py:
```diff
@@ -62,27 +88,18 @@ def main(args=None) -> None:
-    )
+    debug = False
+    if parsed_args.debug:
+        debug = True
```
Yes, good idea - thanks. Commit https://github.com/dls-controls/coniql/pull/44/commits/7cfb3b4406b723e0a4216e74866b5cd0040c7eba.

rjwills28 commented on src/coniql/caplugin.py:
```diff
@@ -162,47 +156,91 @@ def get_display(self) -> Optional[ChannelDisplay]:
+                value.disarm(values.popleft())
+                return maker.channel_from_update(**value.get())
+            except IndexError:
+                # Should not happen but catch in case deque has overflowed
```
Yes you're right - copies over a comment and didn't update it. Changed comment to reference an IndexError if the deque is empty. Commit https://github.com/dls-controls/coniql/pull/44/commits/aed9bd5dfda9f7911981c86e539517bbdfc13588.

rjwills28 commented on src/coniql/caplugin.py:
```diff
@@ -162,47 +156,91 @@ def get_display(self) -> Optional[ChannelDisplay]:
+                meta.disarm(metas.popleft())
+                return maker.channel_from_update(**value.get(), **meta.get())
+            except IndexError:
+                # Should not happen but catch in case deque has overflowed
```
Fixed in commit https://github.com/dls-controls/coniql/pull/44/commits/aed9bd5dfda9f7911981c86e539517bbdfc13588.

rjwills28 commented on src/coniql/app.py:
```diff
@@ -1,49 +1,75 @@
+        execution_context: Optional[ExecutionContext] = None,
+    ) -> None:
+        for error in errors:
+            print(error.message)
```
Updated to be used the `logging` function. Commit: https://github.com/dls-controls/coniql/pull/44/commits/e461f538aa0d1bc05179da20472fd5ff26466e81.

rjwills28 commented on src/coniql/plugin.py:
```diff
@@ -59,38 +48,8 @@ def transport_pv(self, channel_id: str) -> Tuple[str, str]:
-        return device_config
-
-    def plugin_config_id(self, channel_id: str) -> Tuple[Plugin, ChannelConfig, str]:
+    def plugin_config_id(self, channel_id: str) -> Tuple[Plugin, str]:
```
I think installing the Strawberry mypy extension should fix this? Commit: https://github.com/dls-controls/coniql/pull/44/commits/8d2b1171e294ea9d598b1d495d9a346e371a30d9, now includes this.

rjwills28 commented on src/coniql/caplugin.py:
```diff
@@ -214,19 +252,57 @@ async def subscribe_channel(
+                loop = asyncio.get_running_loop()
+            except RuntimeError:
+                # 'RuntimeError: There is no current event loop...'
+                print("No current event loop...")
```
Updated to be used the `logging` function. Commit: https://github.com/dls-controls/coniql/pull/44/commits/e461f538aa0d1bc05179da20472fd5ff26466e81.

rjwills28 commented on src/coniql/caplugin.py:
```diff
@@ -162,47 +156,91 @@ def get_display(self) -> Optional[ChannelDisplay]:
-        # the channel's update_value function should be called.
-        q: asyncio.Queue[Dict[str, AugmentedValue]] = asyncio.Queue()
+    @staticmethod
+    async def __signal_single_channel(value, values, maker, lock):
```
Added docstring and type hints. See commits https://github.com/dls-controls/coniql/pull/44/commits/c88f333c809d4d0cdc95fc7854dc8d007907c852 and https://github.com/dls-controls/coniql/pull/44/commits/6de0004a804d1eac70b65ad8b2480860da2e9e26.

rjwills28 commented on src/coniql/caplugin.py:
```diff
@@ -162,47 +156,91 @@ def get_display(self) -> Optional[ChannelDisplay]:
+
+    @staticmethod
+    async def __signal_double_channel(
+        value, meta, values, metas, maker, value_lock, meta_lock
```
Type hints and docstring added as above.

rjwills28 commented on src/coniql/types.py:
```diff
@@ -84,20 +105,18 @@ def for_number(
         def ndarray_to_base64_array(
             value: np.ndarray, length: int = 0
-        ) -> Optional[Dict[str, str]]:
+        ) -> Optional[Base64Array]:
```
You are correct that this function should always return a Base64Array type. I have removed the Optional[] from the type hint in commit https://github.com/dls-controls/coniql/pull/44/commits/dc65804f9da427d8ad3e4eea6940a1916930365b.

rjwills28 commented on src/coniql/types.py:
```diff
@@ -47,9 +48,29 @@ class ChannelDisplay:
-    form: Optional[DisplayForm], precision: Optional[int]
-) -> str:
+@dataclass
+class NumberType(Enum):
+    INT8 = "INT8"
+    UINT8 = "UINT8"
+    INT16 = "INT16"
+    UINT16 = "UINT16"
+    INT32 = "INT32"
+    UINT32 = "UINT32"
+    INT64 = "INT64"
+    UINT64 = "UINT64"
+    FLOAT32 = "FLOAT32"
+    FLOAT64 = "FLOAT64"
```
I was able to do this when sorting out all of the Strawberry type hinting, see commit: https://github.com/dls-controls/coniql/pull/44/commits/7f137c183012c7b3d79627de8d9beec6a023b423. I was indeed able to stack the decorators.

rjwills28 commented on src/coniql/strawberry_schema.py:
```diff
@@ -0,0 +1,370 @@
+
+
+@strawberry.enum
+class Widget(Enum):
```
I have been able to do this so that I have just one copy that I re-use. See commit: https://github.com/dls-controls/coniql/pull/44/commits/1670d3ae80e85172631e795f042124c087210d4a.

rjwills28 commented on src/coniql/strawberry_schema.py:
```diff
@@ -0,0 +1,370 @@
+
+
+@strawberry.enum
+class DisplayForm(Enum):
```
Also done as above.

rjwills28 commented on tests/test_aiohttp.py:
```diff
@@ -0,0 +1,295 @@
+
+
+@pytest.mark.asyncio
+async def test_subscribe_ticking_graphql_transport_ws_protocol(
```
I have been able to parametrize this test so that we simply have to define the difference between the websocket protocols as inputs to a generic function. See commit https://github.com/dls-controls/coniql/pull/44/commits/caf1cbd0a881cdd95001f7301a8b47c462a7f7e3.
Note however that this has been superseded in my attempts to parametrize both test_caplugin and test_aiohttp tests.

rjwills28 commented on tests/test_caplugin.py:
```diff
@@ -1,355 +1,167 @@
             ),
-        )
+            status=dict(quality="VALID"),
+        ),
```
I did like your suggestion of parametrizing these tests. I have been able to do this although I did come across a few complications parametrizing the different fixtures. I found a way to do this but there are some limitations based on running async fixtures in async tests (see this post: https://github.com/pytest-dev/pytest-asyncio/issues/112). The conclusion is that I have to keep the actual test functions as non-async which forces me to do a lot more of the setup work in the fixtures. 

I have added a new files with the parametrized tests which combines the tests in test_caplugin.py and test_aiohttp.py. See commit https://github.com/dls-controls/coniql/pull/44/commits/83cd7e9405ea29786b1c4e3e55bda8c9dfba0538, new file: tests/test_schema_and_aiohttp.py. I have left the old tests there for the moment as I would like some input as to whether we think the parametrized tests are clearer and easier to work with or not. If we like them then I will remove the old tests and move some of the fixtures from conftest.py that are only used by this new file.

rjwills28 commented:
Thank you for the detailed review. I believe I have now addressed all of the issues. At this point I think the PR should be re-reviewed as there have been some significant changes. 
I would also like some input on the attempt to parametrize some of the tests in commit https://github.com/dls-controls/coniql/pull/44/commits/83cd7e9405ea29786b1c4e3e55bda8c9dfba0538. It removes duplicate code but does add a level of complexity to the writing of tests if you don't understand how it works.

aawdls commented:
> Thank you for the detailed review. I believe I have now addressed all of the issues. At this point I think the PR should be re-reviewed as there have been some significant changes. I would also like some input on the attempt to parametrize some of the tests in commit [83cd7e9](https://github.com/dls-controls/coniql/commit/83cd7e9405ea29786b1c4e3e55bda8c9dfba0538). It removes duplicate code but does add a level of complexity to the writing of tests if you don't understand how it works.

Unfortunately I think that readability of the tests has been lost and I don't think that in its current form it is an improvement.

Can you avoid having all of the one-line fixtures `async def client_get_nan_pv` etc if you split `test_get_pv` into two test functions, one that calls `query_schema` and one that calls `query_aio_httpclient`? Then you would just need to parametrize the query and the result.

Also if parametrizing with async fixtures is complicated, can you consider making some of the input and functions not fixtures but regular data objects and functions? For example, if the functions `client_get_int_pv` etc. were not fixtures, you'd be able to pass the reference to the function into the test instead of having to use `request.getfixturevalue`. In my understanding, fixtures are beneficial when you want to reuse the same returned object; but when you're just using it as a regular function all the extra infrastructure doesn't seem to give a benefit.

aawdls commented on tests/test_schema_and_aiohttp.py:
```diff
@@ -0,0 +1,410 @@
+    return await query_schema(schema, base64_put)
+
+
+int_pv_result = dict(
```
I'd find these easier to read if you used `{}` notation instead of `dict`; is there a reason for this choice?

rjwills28 replied:
I was just following what was there before but I do agree that the `{}` notation is easier to read and also a more efficient way of creating dictionaries. I have updated all dictionary definitions in the tests.

coretl replied:
Probably because I don't like typing `"`, so I tend to write single level dicts as `dict(a=1, b=2, c=3)` rather than `{"a": 1, "b": 2, "c": 3}`. Agree that nested dicts are definitely cleaner with `{}` though.

aawdls commented on src/coniql/app.py:
```diff
@@ -1,49 +1,82 @@
+    ) -> None:
+        for error in errors:
+            if not error.message:
+                logging.error("Unknown error occurred. Enable debugging to find cause.")
```
I would like to suggest a different mechanism for this.

The problem with using the default logging functions `logging.error` is that you are using the root logger rather than a specific one, and you therefore lose the ability to control these log messages with app-level config later. In the base implementation, they use the `StrawberryLogger` class which in turn uses a named logger `strawberry.execution`.

In the base implementation `StrawberryLogger.error` accepts a keyword argument `stack_info` which can be set to False. So one option would be to override this with that argument set to False:

```
    def process_errors(
        self,
        errors: List[GraphQLError],
        execution_context: Optional[ExecutionContext] = None,
    ) -> None:
        for error in errors:
            StrawberryLogger.error(error, execution_context, stack_info=False)
```

Better yet would be to configure logging at the app level in `coniql.app.main`, and define a custom `logging.Formatter` which overrides the https://docs.python.org/3/library/logging.html#logging.Formatter.formatStack formatStack method to suppress the exception information (perhaps conditional on the coniql Debug argument). This would be more resilient to changes of undocumented Strawberry behaviour.

rjwills28 replied:
Thanks for the suggestion and demo code. I have implemented the latter option using a custom logging formatter to suppress the exception stacktrace in commit: https://github.com/dls-controls/coniql/pull/44/commits/61fe51d70f0370cc37ff6819ca21f04d0bab5102. 
This eliminates the need to overwrite the undocumented Strawberry method.

rjwills28 commented:
> Can you avoid having all of the one-line fixtures async def client_get_nan_pv etc if you split test_get_pv into two test functions, one that calls query_schema and one that calls query_aio_httpclient? Then you would just need to parametrize the query and the result.

Yes we could do this but then we will still essentially have a two sets of test, one for the schema and the other for the aiohttp client. I think maybe this is OK though. Trying to make generic functions to allow the parametrization combing these two seems to just make it more confusing. I agree that we could still use parametrization to write all of the 'get' and 'put' tests. I'm inclined to abandon test_schema_and_aiohttp.py and instead refactor test_caplugin.py and test_aiohttp.py this way?

aawdls commented:
> I'm inclined to abandon test_schema_and_aiohttp.py and instead refactor test_caplugin.py and test_aiohttp.py this way?

This seems to me like a good middle ground

AlexanderWells-diamond commented on src/coniql/strawberry_schema.py:
```diff
@@ -0,0 +1,263 @@
+    has type "T", variable has type "Channel"'. Possibly a Strawberry issue
+    as strawberry.subscription is a partial function"""
+
+    subscribeChannel: Optional[Channel] = strawberry.subscription(
+        resolver=subscribe_channel
+    )  # type: ignore
```
More for my own interest than an actual problem, but why is this an `Optional`? I don't understand the need to override the type hinting (and why we suppress mypy's checks) here.

rjwills28 replied:
Not obvious so maybe I should add a comment but when I was testing with multiple machine-status screens that were all unsubscribing and then re-subscribing I noticed that during the disconnect you could get a lot of exceptions as the Channel was not returned as the websocket had been closed. Adding `Optional` to this fixes these errors as it doesn't complain if the Channel isn't returned.

mypy reports the type of the `strawberry.subscription(resolver=...)` as type "T". Looking in the Strawberry code I found that this is actually a partial function. I couldn't find a way to get a type that mypy was happy with so had to add the ignore.

AlexanderWells-diamond commented on src/coniql/strawberry_schema.py:
```diff
@@ -0,0 +1,263 @@
+    ) -> Sequence[Channel]:
+        """Put a list of values to a list of Channels"""
+        store: PluginStore = store_global
+        pvs = []
+        plugins = set()
```
Can these both have type hints added? Especially the set.

rjwills28 replied:
I've added type hints for these.

AlexanderWells-diamond commented on src/coniql/strawberry_schema.py:
```diff
@@ -0,0 +1,263 @@
+        results = []
+        for value in put_values:
+            put_value: Union[str, np.ndarray] = value
+            if value[:1] in "[{":
```
Entirely for my own interest: Under what situations do we receive JSON here?

AlexanderWells-diamond commented on src/coniql/types.py:
```diff
@@ -32,24 +42,68 @@ def contains(self, value: float) -> bool:
+    RW = "RW"
+
+
+@strawberry.type
+@dataclass
```
Is there a reason this class still has the `@dataclass` annotation, but `ChannelDisplay`, above, does not?

AlexanderWells-diamond commented on src/coniql/caplugin.py:
```diff
@@ -59,7 +58,6 @@ def _create_formatter(
         elif isinstance(value, (int, float)):
             # number
             formatter = ChannelFormatter.for_number(
-                display_form,
                 precision,
                 units,
             )
```
Another thing that is strictly speaking not a fault of this PR, but I'm noting it here: The type hints for `for_number` appear to be incorrect - I can't see a way that `precision` and `units` could ever be `None`. The use of `getattr` will mean the values are `0` and `""` respectively, not `None`, so the `Optional` part of the type hint is unnecessary.

This also applies to `for_ndarray`.

rjwills28 replied:
I agree that in this case they cannot be `None`, but in the \_\_init\_\_() of simplugin.py they could be as in the `ChannelDisplay` class both precision and units are `Optional`.

AlexanderWells-diamond reviewed (changes requested):
I noted an issue with module names in setup.cfg that will need fixing before this can be merged. 
Also, as I can't seem to put the comment on the file directly, The `[options.package_data]` section of `setup.cfg` is no longer valid - there are no `*.gql` files left in the project.

Other comments are only minor things, and many are not even directly related to this PR. I leave it to you whether these should be solved here or made into issues. 


Regarding the test parameterization, I don't see why the one-line fixtures were created. They can just be regular functions and called dynamically in the test body. I can't see an advantage to declaring them as fixtures. The parameters can literally be the function reference which gets called in the test body. 
As stated in my original comment on this topic, I do believe that having the same test in two different places is not great. It may be because I'm very used to heavily parameterized tests, but I'm fine with the tests as they are. I can certainly understand that it's much harder to get to grips with though.


I did have an issue running the tests. Multiple runs sometimes showed this error (sometimes in result2, sometimes in result4):

```
tests/test_schema_and_aiohttp.py::test_subscribe_pv[schema_subscribe_ticking-expected_result4] FAILED                                                                                           [100%]

============================================================================================== FAILURES ===============================================================================================
____________________________________________________________________ test_subscribe_pv[schema_subscribe_ticking-expected_result4] _____________________________________________________________________
Traceback (most recent call last):
  File "/home/eyh46967/dev/coniql/tests/test_schema_and_aiohttp.py", line 408, in test_subscribe_pv
    assert len(results) == len(expected_result)
AssertionError: assert 2 == 3
 +  where 2 = len([{'subscribeChannel': {'display': {'precision': 5, 'units': 'mm'}, 'value': {'string': '0.00000 mm'}}}, {'subscribeChannel': {'display': None, 'value': {'string': '1.00000 mm'}}}])
 +  and   3 = len([{'subscribeChannel': {'display': {'precision': 5, 'units': 'mm'}, 'value': {'string': '0.00000 mm'}}}, {'subscribeChannel': {'display': None, 'value': {'string': '1.00000 mm'}}}, {'subscribeChannel': {'display': None, 'value': {'string': '2.00000 mm'}}}])
---------------------------------------------------------------------------------------- Captured log teardown ----------------------------------------------------------------------------------------
ERROR    asyncio:base_events.py:1707 Task was destroyed but it is pending!
task: <Task pending name='Task-907' coro=<<async_generator_athrow without __name__>()>>

```

Also, even if all tests pass, I still see the "Task was destroyed but it is pending!" warning message printed 4 times. Is this intentional? Are some of the tests deliberately trying to kill pending tasks? If not, it's an indicator that there's something wrong in task handling somewhere.

rjwills28 commented:
> Regarding the test parameterization, I don't see why the one-line fixtures were created. They can just be regular functions and called dynamically in the test body. I can't see an advantage to declaring them as fixtures. The parameters can literally be the function reference which gets called in the test body.

The problem I ran into here was that I needed to parametrize the fixtures for the two versions (schema and aiohttp) as one takes the schema directly and the other creates an aiohttp_client. The latter has to be done in a fixture as is a feature of pytest-aiohttp. 
In order to parametrize the fixtures I call them within the tests using the `request.getfixturevalue(fixture_to_run)`. This should all be fine as then I can get my client or schema and pass them into other functions to make the queries and gather results. 

The problem is that the fixture to create the aiohttp_client has to be `async `as we have to `await` the client creation and for some reason there is a problem using `request.getfixturevalue(...)` with an `async` fixture that then means that that test cannot be defined as `async` and so you cannot do any awaits in the test, which are necessary to submit queries and get the results. See similar reports of this issue: https://github.com/pytest-dev/pytest-asyncio/issues/112. I couldn't find or think of any other workaround. The only way I saw was to do all of the async work in the fixtures so that the test didn't need to be async, which is the reason for the one-line fixtures in this case.

rjwills28 commented on Pipfile:
```diff
@@ -11,6 +11,7 @@ coniql = {editable = true, path = "."}
 
 [scripts]
 lint = "pre-commit run --all-files --show-diff-on-failure --color=always -v"
+# Not running tests until full subscription/query/put are implemented
```
Removed comment. Thanks

rjwills28 commented on setup.cfg:
```diff
@@ -63,6 +66,19 @@ console_scripts =
+ignore_missing_imports = True
+
+[mypy-aiohttp_cors.*]
+ignore_missing_imports = True
```
You're right, these aren't required. At some point I was getting errors, which is why I added these but this is not the case now so have removed.

rjwills28 commented on setup.cfg:
```diff
@@ -24,12 +24,13 @@ package_dir =
+    strawberry-graphql
     aioca
     p4p<4.0.0
     ruamel-yaml
```
Thanks for catching this - I have updated this.

AlexanderWells-diamond commented on .github/workflows/docs.yml:
```diff
@@ -26,6 +29,7 @@ jobs:
       - name: Build docs
         uses: dls-controls/pipenv-run-action@v1
         with:
+          python-version: "3.8"
```
I don't see why the matrix is defined above and the python-version is defined here. Usually the `python-version` is defined as `${{ matrix.python }}`, to run it on whatever list is defined in the matrix?

rjwills28 replied:
Yes I agree, it didn't work when I defined it with the `${{ matrix.python }}` definition so I just hardcoded the `3.8` to get this pipeline to pass. Seeing as the pipelines are based on the pipenv installation I decided not to waste too much time getting this perfect as it will need updating as soon as we apply the skeleton anyway.

rjwills28 replied:
I did change the ordering of `with` arguments there so I could try putting it back, but as I said, I was just looking for a quick way to get it to pass.

rjwills28 commented:
> I did have an issue running the tests. Multiple runs sometimes showed this error (sometimes in result2, sometimes in result4):

This occasional failure was coming from the subscription 'ticking' test. I have updated this now to make it more reliable including a wait to ensure we are connected to the websocket before trying to subscribe. 

>Also, even if all tests pass, I still see the "Task was destroyed but it is pending!" warning message printed 4 times. Is this intentional? Are some of the tests deliberately trying to kill pending tasks? If not, it's an indicator that there's something wrong in task handling somewhere.

This was an interesting one. In some cases this was caused by the pytest event loop closing too early, which was fixed by using a new event loop for each 'module' (i.e. each set of tests).
There was another issue though in simplugin.py and test_coniql.py that used it where the simplugin `tasks` were only weakly referenced and so were being GC'ed too early. I introduced a `set` (`task_references`) to track these and clean up afterwards so that they are not GC'ed immediately.
With these two fixes I no longer see these warnings at the end of the tests.

rjwills28 commented:
@aawdls @AlexanderWells-diamond I think I have now finished addressing the second round of review comments. Please feel free to review again or provide feedback on my recent changes.

As a side note, I know we are moving away from the `Pipenv` installation process but I was able to get a fresh install working with it on Python 3.8. To do this I had to update the `setuptools` and `wheel` modules. I updated the Pipefile and Pipefile.lock so this might even now work for you?

AlexanderWells-diamond reviewed (approved):
Re-running the tests and I'm unfortunately still seeing warnings, despite tests passing:
```
======================================================================================= 25 passed in 21.13s =======================================================================================
Task was destroyed but it is pending!
task: <Task pending name='Task-169' coro=<BaseGraphQLTransportWSHandler.handle_connection_init_timeout() running at /home/eyh46967/dev/coniql/venv/lib64/python3.8/site-packages/strawberry/subscriptions/protocols/graphql_transport_ws/handlers.py:82> wait_for=<Future pending cb=[<TaskWakeupMethWrapper object at 0x7f3c0f9452b0>()]>>
Task was destroyed but it is pending!
task: <Task pending name='Task-198' coro=<BaseGraphQLTransportWSHandler.handle_connection_init_timeout() running at /home/eyh46967/dev/coniql/venv/lib64/python3.8/site-packages/strawberry/subscriptions/protocols/graphql_transport_ws/handlers.py:82> wait_for=<Future pending cb=[<TaskWakeupMethWrapper object at 0x7f3c0f7ece50>()]>>
```

This is almost certainly a timing issue, where running the test on different hardware produces the warnings. This issue will likely be exacerbated by GitHub Actions, as some of their runners (especially MacOS) are very slow.
I don't know how much longer it is worth spending on this issue though, as the issue most likely comes from how Pytest is handling itself, which is rather unlike any actual client out there.

I am unable to install this internally using the default `pipenv` command, as several of the modules are not available on the internal PyPi. I haven't spent the time installing the missing ones as we're going to be scrapping `pipenv` anyway.


Other than the test warnings, this all looks good to me.

rjwills28 commented:
I've finally pinpointed where these warnings are coming from and it looks like it's an issue in Strawberry. Essentially they create a task that does an `asyncio.sleep` for 60 seconds. The task then checks whether the `connection_init` message has been received from the client. If so then it simply returns and if not it closes the connection.
In the tests we create a subscription, gather the results and then close the connection and event loop in less than 60 secs. But this task is still running, sitting in the `asyncio.sleep`. It is this task the warning is about. When we run against a 'normal' client we do not have this problem as it is either open long enough for the 60 secs to elapse and complete the task, or we kill the Coniql application. 

This 60 secs is strange to me. In testing I found that this `connection_init` message was received very quickly and so this 60 sec wait seems unnecessary. There are 2 ways to fix it in the Strawberry code:
1. Run a while loop checking the whether the `connection_init` message has been received with a short sleep in between and timeout if we do not receive it after 60 secs. This will cause the task to complete as soon as the `connection_init` message has been received instead of waiting the arbitrary 60 secs.
2. Add some clean up code when a subscription is closed that checks if this task is still running and cancels it if it is.

The second seems cleaner to me but I will have to approach the Strawberry collaboration to get this fixed so might just ask for their opinion.

rjwills28 commented:
Scrap those fixes. Turns out it's much simpler - we can configure the timeout ourselves so could override the 60 seconds. I'll look into that.

rjwills28 commented:
@AlexanderWells-diamond if you have a moment would you mind re-running the tests on your system with my latest changes to see if you still get any warnings? Thanks!

AlexanderWells-diamond commented:
> @AlexanderWells-diamond if you have a moment would you mind re-running the tests on your system with my latest changes to see if you still get any warnings? Thanks!

I've run the tests several times and see no warnings printed.

I'm always a little suspicious about hardcoding any form of timeout into tests. The machines that run the CI have proven to be very slow at times (especially the MacOS ones), which if I understand the changes means that any machine that takes longer than 2 seconds to connect will cause a test failure. I suppose we'll just have to keep an eye on the CI and adjust this timeout accordingly.

rjwills28 commented on src/coniql/app.py:
```diff
@@ -23,16 +24,29 @@ def create_schema(debug: bool):
-        subscription_protocols=[GRAPHQL_TRANSPORT_WS_PROTOCOL, GRAPHQL_WS_PROTOCOL],
-        graphiql=graphiql,
-    )
+    if connection_init_wait_timeout:
+        view = GraphQLView(
+            schema=strawberry_schema,
+            subscription_protocols=[GRAPHQL_TRANSPORT_WS_PROTOCOL, GRAPHQL_WS_PROTOCOL],
+            graphiql=graphiql,
+            connection_init_wait_timeout=connection_init_wait_timeout,
+        )
+    else:  # Use default connection_init_wait_timeout
+        view = GraphQLView(
+            schema=strawberry_schema,
+            subscription_protocols=[GRAPHQL_TRANSPORT_WS_PROTOCOL, GRAPHQL_WS_PROTOCOL],
+            graphiql=graphiql,
```
I did try something like this but I couldn't get it to play nicely with mypy type checking. I was getting errors that 'Argument 4 has incompatible type...'. I'll take a closer look to see it there's a way around this.

rjwills28 replied:
Defining `kwargs: Any = {}` fixes the mypy errors.

rjwills28 replied:
Fixed in latest commit.

rjwills28 commented:
We send the `connection_init` message ourselves in the tests as soon as we have connected (see https://github.com/dls-controls/coniql/blob/ee1304c9115aa21cfd73f3c953d0ef2c9faff21a/tests/test_aiohttp.py#L116) so I would be surprised if we saw a timeout here, however something to keep in mind.
It would be nicer if the Strawberry code cleaned up properly so possibly something we could follow up with them at a later date?

AlexanderWells-diamond commented:
Yes, it sounds like there may be something that could be done in Strawberry so should be raised with them.

Otherwise this all looks good.

aawdls commented:
This is now ready to be merged and will be deployed in pre-prod for further evaluation at scale.
